### PR TITLE
Avoid adding empty splices in `CodedBufferWriter`

### PR DESCRIPTION
**Pull request by osa1**
By avoiding allocating empty `Uint8List`s this saves a few percent in dart2js -O4 in the benchmark reported in #885.

|                              | Before     | After      | Diff                |
|------------------------------|------------|------------|---------------------|
| AOT                          | 122,917 us | 122,741 us |                     |
| JIT                          |  93,376 us |  94,880 us |                     |
| dart2js -O4                  | 271,111 us | 258,250 us | -12,861 us, -4.7%   |
| dart2wasm --omit-type-checks | 196,454 us | 195,300 us |                     |

Number of splices in the benchmark before this change: 21,123
After: 20,857

### CODE REVIEW

File: protobuf/lib/src/protobuf/coded_buffer_writer.dart [hunk 1]
```diff
@@ -337,11 +337,22 @@ class CodedBufferWriter {
         _writeVarint32(value ? 1 : 0);
         break;
       case PbFieldType._BYTES_BIT:
-        _writeBytesNoTag(
-            value is Uint8List ? value : Uint8List.fromList(value));
+        final List<int> bytes = value;
+        if (bytes is Uint8List) {
+          _writeBytesNoTag(bytes);
```
Comment (mkustermann) [inline protobuf/lib/src/protobuf/coded_buffer_writer.dart:342]: If `bytes.length == 0` we still add a splice, etc. Consider adding a `bytes.length == 0` case here inside the `if (is Uint8List)` (as that `.length` load will be inlined if we know it's a `Uint8List`)

File: protobuf/lib/src/protobuf/coded_buffer_writer.dart [hunk 2]
```diff
@@ -337,11 +337,22 @@ class CodedBufferWriter {
...
+        } else if (bytes.isEmpty) {
+          writeInt32NoTag(0);
```
Comment (mkustermann) [inline protobuf/lib/src/protobuf/coded_buffer_writer.dart:344]: Consider introducing a `_writeEmptyBytes()` helper and use it here and other places where we basically intend to write an empty byte array.

```
@pramga('vm:prefer-inline')
@...
void _writeEmptyBytes() {
  writeInt32NoTag(0);
}
```

**Review by mkustermann — APPROVED** [review]

**Comment by osa1** [comment]
I will get back to this after cl/576427817.

File: protobuf/lib/src/protobuf/coded_buffer_writer.dart (hunk 2 above)
Comment (osa1) [reply protobuf/lib/src/protobuf/coded_buffer_writer.dart]: I added the helper but I didn't add the pragmas. We have a bunch of other one-liners in this file, I'd like to experiment with inlining them manually separately.

**Review by mkustermann — APPROVED** [review]

**Pull request merged**